atom/browser/api/atom_api_notification.cc

```diff
@@ -0,0 +1,199 @@
+  InitWith(isolate, wrapper);
+
+  mate::Dictionary opts;
+  if (args->GetNext(&opts)) {
+    opts.Get("title", &title_);
+    opts.Get("body", &body_);
+    has_icon_ = opts.Get("icon", &icon_);
+    if (has_icon_) {
+      opts.Get("icon", &icon_path_);
+    }
+    opts.Get("silent", &silent_);
+    opts.Get("replyPlaceholder", &reply_placeholder_);
+    opts.Get("hasReply", &has_reply_);
+    id_ = id_counter++;
+  }
+  notifications_[id_] = this;
+  OnInitialProps();
+}
+
+Notification::~Notification() {}
+
+// static
+mate::WrappableBase* Notification::New(mate::Arguments* args) {
+  if (!Browser::Get()->is_ready()) {
+    args->ThrowError("Cannot create Notification before app is ready");
+    return nullptr;
+  }
+  return new Notification(args->isolate(), args->GetThis(), args);
+}
+
+bool Notification::HasID(int id) {
+  return notifications_.find(id) != notifications_.end();
+}
+
+Notification* Notification::FromID(int id) {
+  return notifications_[id];
+}
+
+// Getters
+int Notification::GetID() {
+  return id_;
+}
+base::string16 Notification::GetTitle() {
+  return title_;
+}
+base::string16 Notification::GetBody() {
+  return body_;
+}
+bool Notification::GetSilent() {
+  return silent_;
+}
+base::string16 Notification::GetReplyPlaceholder() {
+  return reply_placeholder_;
+}
+bool Notification::GetHasReply() {
+  return has_reply_;
+}
+
+// Setters
+void Notification::SetTitle(base::string16 new_title) {
+  title_ = new_title;
+  NotifyPropsUpdated();
+}
+void Notification::SetBody(base::string16 new_body) {
+  body_ = new_body;
+  NotifyPropsUpdated();
+}
+void Notification::SetSilent(bool new_silent) {
+  silent_ = new_silent;
+  NotifyPropsUpdated();
+}
+void Notification::SetReplyPlaceholder(base::string16 new_reply_placeholder) {
+  reply_placeholder_ = new_reply_placeholder;
+  NotifyPropsUpdated();
+}
+void Notification::SetHasReply(bool new_has_reply) {
+  has_reply_ = new_has_reply;
+  NotifyPropsUpdated();
+}
+
+void Notification::OnClicked() {
+  Emit("click");
+}
+
+void Notification::OnReplied(std::string reply) {
+  Emit("reply", reply);
+}
+
+void Notification::OnShown() {
+  Emit("show");
+}
+
+void Notification::OnClosed() {
+  Emit("close");
+}
+
+void Notification::NotifyPropsUpdated() {}
+
+// Showing notifications
+void Notification::Show() {
+  SkBitmap image = *(new SkBitmap);
+  if (has_icon_) {
+    image = *(icon_.ToSkBitmap());
+  }
+
+  std::unique_ptr<AtomNotificationDelegateAdapter> adapter(
+      new AtomNotificationDelegateAdapter(this));
+  auto notif = presenter_->CreateNotification(adapter.get());
+  if (notif) {
+    ignore_result(adapter.release());  // it will release itself automatically.
+    GURL nullUrl = *(new GURL);
+    notif->Show(title_, body_, "", nullUrl, image, silent_, has_reply_, reply_placeholder_);
+  }
+}
+
+bool initialized_ = false;
+void Notification::OnInitialProps() {
+  if (!initialized_) {
+    presenter_ = brightray::NotificationPresenter::Create();
```

Comment:
The presenter can be received by calling `BrowserClient::GetNotificationPresenter()`.

Also it should be noticed that not every platform supports notifications, so it might return null and this case should be guarded. We also need to add a new API like `Notification.isSupported()` to indicate whether notification is supported on current platform.